core/extension/extension_api_dump.cpp

```diff
@@ -0,0 +1,510 @@
+						arguments.push_back(d3);
+					}
+					if (arguments.size()) {
+						d2["arguments"] = arguments;
+					}
+					constructors.push_back(d2);
+				}
+
+				if (constructors.size()) {
+					d["constructors"] = constructors;
+				}
+			}
+
+			builtins.push_back(d);
+		}
+
+		api_dump["builtin_classes"] = builtins;
+	}
+
+	{
+		// classes
+		Array classes;
+
+		List<StringName> class_list;
+
+		ClassDB::get_class_list(&class_list);
+
+		class_list.sort_custom<StringName::AlphCompare>();
+
+		for (List<StringName>::Element *E = class_list.front(); E; E = E->next()) {
+			Dictionary d;
+			StringName class_name = E->get();
+			d["name"] = String(class_name);
+			d["is_refcounted"] = ClassDB::is_parent_class(class_name, "RefCounted");
+			d["is_instantiable"] = ClassDB::can_instantiate(class_name);
+			StringName parent_class = ClassDB::get_parent_class(class_name);
+			if (parent_class != StringName()) {
+				d["inherits"] = String(parent_class);
+			}
+
+			{
+				ClassDB::APIType api = ClassDB::get_api_type(class_name);
+				static const char *api_type[5] = { "core", "editor", "extension", "editor_extension" };
+				d["api_type"] = api_type[api];
+			}
+
+			{
+				//constants
+				Array constants;
+				List<String> constant_list;
+				ClassDB::get_integer_constant_list(class_name, &constant_list, true);
+				for (List<String>::Element *F = constant_list.front(); F; F = F->next()) {
+					StringName enum_name = ClassDB::get_integer_constant_enum(class_name, F->get());
+					if (enum_name != StringName()) {
+						continue; //enums will be handled on their own
+					}
+
+					Dictionary d2;
+					d2["name"] = String(F->get());
+					d2["value"] = ClassDB::get_integer_constant(class_name, F->get());
+
+					constants.push_back(d2);
+				}
+
+				if (constants.size()) {
+					d["constants"] = constants;
+				}
+			}
+			{
+				//enum
+				Array enums;
+				List<StringName> enum_list;
+				ClassDB::get_enum_list(class_name, &enum_list, true);
+				for (List<StringName>::Element *F = enum_list.front(); F; F = F->next()) {
+					Dictionary d2;
+					d2["name"] = String(F->get());
+
+					Array values;
+					List<StringName> enum_constant_list;
+					ClassDB::get_enum_constants(class_name, F->get(), &enum_constant_list, true);
+					for (List<StringName>::Element *G = enum_constant_list.front(); G; G = G->next()) {
+						Dictionary d3;
+						d3["name"] = String(G->get());
+						d3["value"] = ClassDB::get_integer_constant(class_name, G->get());
+						values.push_back(d3);
+					}
+
+					d2["values"] = values;
+
+					enums.push_back(d2);
+				}
+
+				if (enums.size()) {
+					d["enums"] = enums;
+				}
+			}
+			{
+				//methods
+				Array methods;
+				List<MethodInfo> method_list;
+				ClassDB::get_method_list(class_name, &method_list, true);
+				for (List<MethodInfo>::Element *F = method_list.front(); F; F = F->next()) {
+					if (F->get().name.begins_with("_")) {
+						continue; //hidden method
```

Comment:
Methods starting with "_" are also for virtual functions like "_entre_tree", "_process", "_draw", not necessary private methods.

Reply:
Yeah, Godot core should have API for marking methods as internal explicitly rather than relying on `_`, which can mean both "internal/private" and "virtual", so this looks error-prone to me. See also godotengine/godot-proposals#2285.

Reply:
By curiosity, why do we register those methods in the ClassDB in the first place if they are not meant to be exposed?

Reply:
My mistake, I need to check whether they are virtual and expose them.